eth/catalyst/queue.go

```diff
@@ -34,11 +35,47 @@ const maxTrackedPayloads = 10
 // latest one; but have a slight wiggle room for non-ideal conditions.
 const maxTrackedHeaders = 10
 
+// payload wraps the execution payload generated by miner
+type payload struct {
+	lock   sync.Mutex
+	done   bool
+	empty  *types.Block
+	block  *types.Block
+	result chan *types.Block
+}
+
+// resolve extracts the generated full block from the given channel if possible
+// or fallback to empty block as an alternative.
+func (req *payload) resolve() *beacon.ExecutableDataV1 {
+	// this function can be called concurrently, prevent any
+	// concurrency issue in the first place.
+	req.lock.Lock()
+	defer req.lock.Unlock()
+
+	// Try to resolve the full block first if it's not obtained
+	// yet. The returned block can be nil if the generation fails.
+	if !req.done {
+		select {
+		case req.block = <-req.result:
+			req.done = true
+		case <-time.NewTimer(time.Millisecond * 100).C:
+			// TODO(rjl49345642, Marius), should we keep this
+			// 100ms timeout allowance? Why not just use the
+			// default and then fallback to empty directly?
```

Comment:
Ah, also, new time used like this will leak goroutines. You should do it prior to select:

```
timeout := time.NewTimer(500 * time.Millisecond)
defer timeout.Stop()

select {
...
}
```

Reply:
Also always use the number first and then `* time.Millisecond`. The reverse way used above reads weird.

Reply:
Thats the asian way, from left to right